src/lib/Inline.hs

```diff
@@ -0,0 +1,497 @@
+
+inlineName :: Emits o => Context SExpr e o -> SAtomName i -> InlineM i o (e o)
+inlineName ctx name =
+  lookupSubstM name >>= \case
+    Rename name' -> do
+      -- This is the considerInline function from the Secrets paper; this
+      -- is where we decide whether to inline a binding that isn't to be
+      -- inlined unconditionally.
+      -- For now, we just don't.  If we did, we would want to start with
+      -- something like
+      -- lookupEnv name' >>= \case
+      --   (expr', presInfo) | inline presInfo expr' ctx -> inline
+      --   no info -> do not inline (as now)
+      reconstruct ctx (Atom $ Var name')
+    SubstVal (DoneEx expr) -> dropSubst $ inlineExpr ctx expr
+    SubstVal (SuspEx expr s') -> withSubst s' $ inlineExpr ctx expr
+
+instance Inlinable SAtomName where
+  inline ctx a = inlineName (EmitToAtomCtx $ EmitToNameCtx ctx) a
+
+instance Inlinable SAtom where
+  inline ctx a = inlineAtom (EmitToAtomCtx ctx) a
+
+instance Inlinable SBlock where
+  inline ctx (Block ann decls ans) = case (ann, decls) of
+    (NoBlockAnn, Empty) ->
+      (Block NoBlockAnn Empty <$> inline Stop ans) >>= reconstruct ctx
+    (NoBlockAnn, _) -> error "should be unreachable"
+    (BlockAnn ty effs, _) -> do
+      (Abs decls' ans') <- buildScoped $ inlineDecls decls $ inline Stop ans
+      ty' <- inline Stop ty
+      effs' <- inline Stop effs  -- TODO Really?
+      reconstruct ctx $ Block (BlockAnn ty' effs') decls' ans'
+
+-- Still using InlineM because we may call back into inlining, and we wish to
+-- retain our output binding environment.
+reconstruct :: Emits o => Context e1 e2 o -> e1 o -> InlineM i o (e2 o)
+reconstruct ctx e = case ctx of
+  Stop -> return e
+  TabAppCtx ixs s ctx' -> withSubst s $ reconstructTabApp ctx' e ixs
+  EmitToAtomCtx ctx' -> emitExprToAtom e >>= reconstruct ctx'
+  EmitToNameCtx ctx' -> emitAtomToName noHint e >>= reconstruct ctx'
+  RepWrap ctx' -> reconstruct ctx' $ toE e
+{-# INLINE reconstruct #-}
+
+reconstructTabApp :: Emits o
+  => Context SExpr e o -> SExpr o -> NonEmpty (SAtom i) -> InlineM i o (e o)
+reconstructTabApp ctx expr ixs =
+  case fromNaryForExpr (NE.length ixs) expr of
+    Just (bsCount, NaryLamExpr bs _ (Block _ decls result)) -> do
+      let (ixsPref, ixsRest) = NE.splitAt bsCount ixs
+      -- Note: There's a decision here.  Is it ok to inline the atoms in
+      -- `ixsPref` into the body `decls`?  If so, should we pre-process them and
+      -- carry them in `DoneEx`, or suspend them in `SuspEx`?  (If not, we can
+      -- emit fresh bindings and use `Rename`.)  This decision affects the
+      -- runtime of the inliner and the code size of the IR the inliner
+      -- produces; since `ixsPref` are atoms, no decision made here can
+      -- duplicate work.
+      -- - Tried going with `SuspEx`, in the interest of "launch and iterate".
+      -- - This creates a bug: what if I'm beta-reducing a body that
```

Comment:
This sounds like you've exposed a preëxisting bug rather than created a new one. Beta-reducing a `for` expression with a correctly-typed atom shouldn't crash the compiler. We do it all over the place. In particular, `j:(..<some_atom>)` should be perfectly fine. I'm curious what the error message was, if you can remember it.

Reply:
On second thought, this is more subtle.  Each `ixsPref` is an atom, but there may a pending inlining that will try to turn it into an expression (that may do work).  So my original tri-lemma is not actually valid at all; rewrote the comment.